Meter.c

```diff
@@ -266,89 +317,610 @@ static void BarMeterMode_draw(Meter* this, int x, int y, int w) {
+      double nextThresholdHigh = 0.0;
+      double nextThresholdLow = 1.0;
+      int nextCellLimitHigh = 0;
+      int nextCellLimitLow = topCell;
+      uint8_t numTiedItems = 0;
+      for (uint8_t i = 0; i <= topCellItem && sum < DBL_MAX; i++) {
+         if (!(this->values[i] > 0.0)) // Skip negative and NaN values
+            continue;
+         sum += this->values[i];
+         if (!(sum <= DBL_MAX))
+            sum = DBL_MAX;
+
+         double top = (sum / scaledTotal) * GRAPH_HEIGHT;
+         double area = top - bottom;
+         double remainder = area;
+         if (i == topCellItem) {
+            remainder -= topCellArea;
+            if (!(remainder >= 0.0)) {
+               remainder = 0.0;
+            }
+         }
+         int numCells = (int)remainder;
+         remainder -= numCells;
+
+         // Whether the item will receive an extra cell or be truncated
+         if (remainder >= threshold) {
+            if (remainder > threshold) {
+               if (remainder < nextThresholdLow) {
+                  nextThresholdLow = remainder;
+               }
+               numCells++;
+            } else if (numCells <= cellLimit) {
+               if (numCells < cellLimit) {
+                  if (numCells > nextCellLimitHigh) {
+                     nextCellLimitHigh = numCells;
+                  }
+                  numCells++;
+               } else {
+                  numTiedItems++;
+                  if (numTiedItems <= tiedItemLimit) {
+                     numCells++;
+                  } else {
+                     remainder = 0.0;
+                  }
+               }
+            } else {
+               if (numCells < nextCellLimitLow) {
+                  nextCellLimitLow = numCells;
+               }
+               remainder = 0.0;
+            }
+         } else {
+            if (remainder > nextThresholdHigh) {
+               nextThresholdHigh = remainder;
+            }
+            remainder = 0.0;
+         }
 
-#endif
+         // Paint cells to the buffer
+         uint8_t blanksAtEnd = 0;
+         if (i == topCellItem && topCellArea > 0.0) {
+            numCells++;
+            if (area < topCellArea) {
+               remainder = MAXIMUM(area, 0.25);
+               blanksAtEnd = (uint8_t)blanksAtTopCell;
+            }
+         } else if (cellsPainted + numCells >= topCell) {
+            blanksAtEnd = 0;
+         } else if (cellsPainted <= 0 || bottom <= cellsPainted) {
+            blanksAtEnd = ((uint8_t)((1.0 - remainder) * 8.0) % 8);
+         } else if (cellsPainted + numCells > top) {
+            assert(cellsPainted + numCells - top < 1.0);
+            blanksAtEnd = (uint8_t)((cellsPainted + numCells - top) * 8.0);
+         }
+         unsigned int blanksAtStart = 0;
+         if (cellsPainted > 0) {
+            blanksAtStart = ((uint8_t)((1.0 - remainder) * 8.0) % 8 - blanksAtEnd);
+         }
+         while (numCells > 0 && cellsPainted < topCell) {
```

Comment:
Some more visual grouping of operations
```suggestion

         unsigned int blanksAtStart = 0;
         if (cellsPainted > 0) {
            blanksAtStart = ((uint8_t)((1.0 - remainder) * 8.0) % 8 - blanksAtEnd);
         }

         while (numCells > 0 && cellsPainted < topCell) {
```